File: 3.5/drawNURBSSurf.m
%
% Project: Approximation and Finite Elements in Isogeometric Problems
% Author:  Luca Carlon
% Date:    2009.10.25
%
% Copyright (c) 2009-2021 Luca Carlon. All rights reserved.
%
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program. If not, see <http://www.gnu.org/licenses/>.
%

function [] = drawNURBSSurf(n, p, Xi, m, q, Eta, Pw)

% Draw the physical domain.
hold on;
axis equal;
box on;
grid on;
xlabel('x');
ylabel('y');
zlabel('z');

% Define the sample values.
spacexi = linspace(0, 1, 100);
spaceeta = linspace(0, 1, 100);

% Define the points to be drawn.
for i = 1:length(spacexi)
    for j = 1:length(spaceeta)
        S = computeNURBSSurfPoint(n, p, Xi, m, q, Eta, Pw, spacexi(i), spaceeta(j));
        X(i, j) = S(1);
        Y(i, j) = S(2);
        Z(i, j) = S(3);
    end
end
surfc(X, Y, Z);
shading flat;
clear S;

% Draw the control points.
for i = 1:length(Pw(:, 1, 1))
    for j = 1:length(Pw(1, :, 1))
        plot3(Pw(i, j, 1)./Pw(i, j, 4), Pw(i, j, 2)./Pw(i, j, 4),...
            Pw(i, j, 3)./Pw(i, j, 4),...
            '.', 'MarkerSize', 10, 'Color', 'red');
    end
end

% Draw the lines.
for i = 1:length(Pw(:, 1, 1))-1
    for j = 1:length(Pw(1, :, 1))-1
        plot3([Pw(i, j, 1)./Pw(i, j, 4), Pw(i+1, j, 1)./Pw(i+1, j, 4)],...
            [Pw(i, j, 2)./Pw(i, j, 4), Pw(i+1, j, 2)./Pw(i+1, j, 4)],...
            [Pw(i, j, 3)./Pw(i, j, 4), Pw(i+1, j, 3)./Pw(i+1, j, 4)],...
            '--');
        plot3([Pw(i, j, 1)./Pw(i, j, 4), Pw(i, j+1, 1)./Pw(i, j+1, 4)],...
            [Pw(i, j, 2)./Pw(i, j, 4), Pw(i, j+1, 2)./Pw(i, j+1, 4)],...
            [Pw(i, j, 3)./Pw(i, j, 4), Pw(i, j+1, 3)./Pw(i, j+1, 4)],...
            '--');
    end
end

% Draw the isocurves on Eta.
for j = 1:length(Eta)
    for i = 1:length(spacexi)
        S(i, :) = computeNURBSSurfPoint(n, p, Xi, m, q, Eta, Pw, spacexi(i), Eta(j));
    end
    plot3(S(:, 1), S(:, 2), S(:, 3), 'k');
end

% Draw the isocurves on Xi.
for i = 1:length(Xi)
    for j = 1:length(spaceeta)
        S(j, :) = computeNURBSSurfPoint(n, p, Xi, m, q, Eta, Pw, Xi(i), spaceeta(j));
    end
    plot3(S(:, 1), S(:, 2), S(:, 3), 'k');
end

File: ts/src/nurbs/drawNurbsSurf.ts
/**
 * Project: Approximation and Finite Elements in Isogeometric Problems
 * Author:  Luca Carlon
 * Date:    2021.06.10
 *
 * Copyright (c) 2021 Luca Carlon. All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { Matrix2, RowVector } from "../core/matrix"
import { Point } from "../core/point"
import { NurbsSurf } from "./nurbs"

/**
 * Draws a NURBS surface.
 * 
 * @param controlPoints 
 * @param Xi 
 * @param Eta 
 * @param w 
 * @param p 
 * @param q 
 * @param drawControlPoints 
 * @param plot 
 */
export let drawNurbsSurf = (
    controlPoints: Point[][],
    Xi: RowVector,
    Eta: RowVector,
    w: Matrix2,
    p: number,
    q: number,
    drawControlPoints: boolean,
    plot: string,
    sameScale: boolean = false,
    maxXi = 1,
    maxEta = 1,
    title: string = "NURBS surface") => {
    const nurbs = new NurbsSurf(controlPoints, Xi.toArray(), Eta.toArray(), w, p, q)
    // @ts-expect-error
    const xiValues = math.range(0, maxXi, 0.005*maxXi).toArray()
    // @ts-expect-error
    const etaValues = math.range(0, maxEta, 0.005*maxEta).toArray()
    let xValues = []
    let yValues = []
    let zValues = []
    xiValues.map((xi: number) => {
        etaValues.map((eta: number) => {
            let p = nurbs.evaluate(xi, eta)
            xValues.push(p.x())
            yValues.push(p.y())
            zValues.push(p.z())
        })
    })
    let min = zValues.reduce(function (a, b) {
        return Math.min(a, b);
    });
    let max = zValues.reduce(function (a, b) {
        return Math.max(a, b);
    });
    const trace1 = {
        x: xValues,
        y: yValues,
        z: zValues,
        mode: "markers",
        type: "scatter3d",
        marker: {
            size: 1,
            color: zValues.slice(0).sort(),
            colorscale: "Jet",
            cmin: min,
            cmax: max
        },
        line: {}
    }
    const data: any[] = [trace1]

    if (drawControlPoints) {
        for (let cpxs of controlPoints) {
            const cpXValues = []
            const cpYValues = []
            const cpZValues = []
            for (let controlPoint of cpxs) {
                cpXValues.push(controlPoint.x())
                cpYValues.push(controlPoint.y())
                cpZValues.push(controlPoint.z())
            }

            const trace = {
                x: cpXValues,
                y: cpYValues,
                z: cpZValues,
                name: "Control points",
                mode: 'lines',
                type: "scatter3d",
                line: {
                    color: "red",
                    width: 3,
                    dash: "dashdot"
                }
            }
            data.push(trace)
        }

        for (let j = 0; j < controlPoints[0].length; j++) {
            const cpXValues = []
            const cpYValues = []
            const cpZValues = []
            for (let i = 0; i < controlPoints.length; i++) {
                cpXValues.push(controlPoints[i][j].x())
                cpYValues.push(controlPoints[i][j].y())
                cpZValues.push(controlPoints[i][j].z())

                const trace = {
                    x: cpXValues,
                    y: cpYValues,
                    z: cpZValues,
                    name: "Control points",
                    mode: 'lines+markers',
                    type: "scatter3d",
                    line: {
                        color: "red",
                        width: 3,
                        dash: "dashdot"
                    },
                    marker: {
                        color: "transparent",
                        size: 8,
                        line: {
                            color: "black",
                            width: 2
                        }
                    }
                }
                data.push(trace)
            }
        }
    }

    var layout = {
        title: {
            text: title,
            font: {
                family: "Ubuntu",
                size: 24,
            }
        },
        height: 700,
        showlegend: false,
        scene: {
            aspectmode: "data",
            aspectratio: {
                x: 1,
                y: 1,
                z: 1
            },
            yaxis: {
                title: {
                    text: "y",
                    font: {
                        family: "Ubuntu",
                        size: 18,
                        color: "#7f7f7f"
                    },
                }
            },
            zaxis: {
                title: {
                    text: "z",
                    font: {
                        family: "Ubuntu",
                        size: 18,
                        color: "#7f7f7f"
                    }
                }
            }
        },
        margin: {
            l: 0,
            r: 0,
            b: 0
        }
    };
    // @ts-expect-error
    Plotly.newPlot(plot, data, layout)
}